File: CSharpCodeBase/components/controlsystem/actionmachine.lua
local Class = require 'utils.hump.class'

local ActionMachine = Class({init=function (self, emptyAction)
      self.buffer = {}
      self._bufferLimit = 2
      self._previousAction = nil
      self._emptyAction = emptyAction
end})

function ActionMachine:GetCurrentAction()
  return #self.buffer > 0 and self.buffer[#self.buffer] or self._emptyAction
end

function ActionMachine:Flush()
  local currentAction = self:GetCurrentAction()
  for k, v in pairs(self.buffer) do
    v:OnRemove()
    v:Free()
  end
  self.buffer = {}
end

function ActionMachine:FixedUpdate()
  local currentAction = self:GetCurrentAction()
  if currentAction ~= nil and currentAction:IsStarted() then
    currentAction:FixedUpdate()
  end
end

function ActionMachine:Update()
  local currentAction = self:GetCurrentAction()
  if currentAction ~= self._previousAction then
    if self._previousAction ~= nil and self._previousAction:IsStarted() then
      self._previousAction:OnSuspend()
    end
    if currentAction ~= nil then
      if currentAction:IsStarted() then
        currentAction:OnResume()
      else
        currentAction:OnStart()
      end
    end
    self._previousAction = currentAction
  elseif currentAction ~= nil and currentAction == self._previousAction and not currentAction:IsStarted() then
    currentAction:OnStart()
  end
  
  if currentAction ~= nil then
    if currentAction:Update() then
      table.remove(self.buffer, #self.buffer)
      currentAction:OnComplete()
      currentAction:Free()
      self._previousAction = nil
    end
  end
end

function ActionMachine:IsActionExists(name)
  for k, v in pairs(self.buffer) do
    if v.name == name then return true end
  end
  return false
end

function ActionMachine:InsertAction(action)
  local idx = 1
  for k, v in pairs(self.buffer) do
    if v:GetPriority() <= action:GetPriority() then
      idx = idx + 1
    else 
      break
    end
  end
  table.insert(self.buffer, idx, action)  
  return idx
end

function ActionMachine:CutLimit()
  while #self.buffer > self._bufferLimit do
    self.buffer[1]:OnRemove()
    self.buffer[1]:Free()
    table.remove(self.buffer, 1)
  end
end

function ActionMachine:RemoveActionsWithSamePriority()
  local previousPriority = self.buffer[#self.buffer]:GetPriority()
  for i = #self.buffer - 1, 1, -1 do
    local currentPriority = self.buffer[i]:GetPriority()
    if currentPriority == previousPriority then
      self.buffer[i]:OnRemove()
      self.buffer[i]:Free()
      table.remove(self.buffer, i)
    else
      previousPriority = currentPriority
    end
  end
end

function ActionMachine:RemoveUncontinuousActions()
  local needIterate = true
  while needIterate do
    local isRemoved = false
    for k, v in pairs(self.buffer) do
      if k < #self.buffer then
        if not v:IsContinuous() and v:IsStarted() then
          v:OnRemove()
          v:Free()
          table.remove(self.buffer, k)
          isRemoved = true
          break
        end
      end
      needIterate = isRemoved
    end
  end
end

function ActionMachine:PushAction(action)
  local currentAction = self:GetCurrentAction()
  self:InsertAction(action)
  self:CutLimit()
  self:RemoveActionsWithSamePriority()
  self:RemoveUncontinuousActions()
  action:OnPushed()
end

function ActionMachine:IsEmpty()
  return #self.buffer == 0
end
return ActionMachine

File: CSharpCodeBase/components/controlsystem/actionmachine.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class actionmachine {
 public void init (self, emptyAction);
       this.buffer = {}
       this._bufferLimit = 2;
       this._previousAction = null;
       this._emptyAction = emptyAction;
 }})
 public void GetCurrentAction(){
   return #this.buffer > 0  &&  this.buffer[#this.buffer]  ||  this._emptyAction;
 }
 public void Flush(){
   var currentAction = self:GetCurrentAction();
   for(k, v in pairs(this.buffer) ){
     v:OnRemove();
     v:Free();
   }
   this.buffer = {}
 }
 public void FixedUpdate(){
   var currentAction = self:GetCurrentAction();
   if(currentAction != null  &&  currentAction:IsStarted()  ){ 
     currentAction:FixedUpdate();
   }
 }
 public void Update(){
   var currentAction = self:GetCurrentAction();
   if(currentAction != this._previousAction  ){ 
     if(this._previousAction != null  &&  this._previousAction:IsStarted()  ){ 
       this._previousAction:OnSuspend();
     }
     if(currentAction != null  ){ 
       if(currentAction:IsStarted()  ){ 
         currentAction:OnResume();
       }else{
         currentAction:OnStart();
       }
     }
     this._previousAction = currentAction;
   }else{if currentAction != null  &&  currentAction == this._previousAction  &&  not currentAction:IsStarted()  ){ 
     currentAction:OnStart();
   }
   
   if(currentAction != null  ){ 
     if(currentAction:Update()  ){ 
       table.remove(this.buffer, #this.buffer);
       currentAction:OnComplete();
       currentAction:Free();
       this._previousAction = null;
     }
   }
 }
 public void IsActionExists(name){
   for(k, v in pairs(this.buffer) ){
     if(v.name == name  ){  return true }
   }
   return false;
 }
 public void InsertAction(action){
   var idx = 1;
   for(k, v in pairs(this.buffer) ){
     if(v:GetPriority() <= action:GetPriority()  ){ 
       idx = idx + 1;
     }else{ 
       break;
     }
   }
   table.insert(this.buffer, idx, action)  ;
   return idx;
 }
 public void CutLimit(){
   while(#this.buffer > this._bufferLimit ){
     this.buffer[1]:OnRemove();
     this.buffer[1]:Free();
     table.remove(this.buffer, 1);
   }
 }
 public void RemoveActionsWithSamePriority(){
   var previousPriority = this.buffer[#this.buffer]:GetPriority();
   for(i = #this.buffer - 1, 1, -1 ){
     var currentPriority = this.buffer[i]:GetPriority();
     if(currentPriority == previousPriority  ){ 
       this.buffer[i]:OnRemove();
       this.buffer[i]:Free();
       table.remove(this.buffer, i);
     }else{
       previousPriority = currentPriority;
     }
   }
 }
 public void RemoveUncontinuousActions(){
   var needIterate = true;
   while(needIterate ){
     var isRemoved = false;
     for(k, v in pairs(this.buffer) ){
       if(k < #this.buffer  ){ 
         if(not v:IsContinuous()  &&  v:IsStarted()  ){ 
           v:OnRemove();
           v:Free();
           table.remove(this.buffer, k);
           isRemoved = true;
           break;
         }
       }
       needIterate = isRemoved;
     }
   }
 }
 public void PushAction(action){
   var currentAction = self:GetCurrentAction();
   self:InsertAction(action);
   self:CutLimit();
   self:RemoveActionsWithSamePriority();
   self:RemoveUncontinuousActions();
   action:OnPushed();
 }
 public void IsEmpty(){
   return #this.buffer == 0;
 }
 }}